## Import Libraries & Load Dataset

In [1]:
import pandas as pd
import numpy as np

titanic = pd.read_csv("titanic (1).csv")
titanic.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


## Explore Dataset

In [2]:
print("Shape:")
print(titanic.shape)

print("\nInfo:")
titanic.info()

print("\nDescribe:")
print(titanic.describe(include="all"))

print("\nMissing Values:")
print(titanic.isnull().sum())

Shape:
(891, 12)

Info:
<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB

Describe:
        PassengerId    Survived      Pclass                     Name   Sex  \
count    891.000000  891.000000  891.000000                      891   891   
unique          NaN         NaN         NaN                      891

## Select Columns

In [3]:
selected_df = titanic[["Survived", "Pclass", "Sex", "Age", "Fare"]]

selected_df.head()

,Survived,Pclass,Sex,Age,Fare
0,0,3,male,22.0,7.2500
1,1,1,female,38.0,71.2833
2,1,3,female,26.0,7.9250
3,1,1,female,35.0,53.1000
4,0,3,male,35.0,8.0500


## Filter Female, First Class, Survived

In [4]:
female_first_survived = titanic[
    (titanic["Sex"] == "female") &
    (titanic["Pclass"] == 1) &
    (titanic["Survived"] == 1)
]

female_first_survived

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
11,12,1,1,"Bonnell, Miss. Elizabeth",female,58.0,0,0,113783,26.5500,C103,S
31,32,1,1,"Spencer, Mrs. William Augustus (Marie Eugenie)",female,NaN,1,0,PC 17569,146.5208,B78,C
52,53,1,1,"Harper, Mrs. Henry Sleeper (Myna Haxtun)",female,49.0,1,0,PC 17572,76.7292,D33,C
...,...,...,...,...,...,...,...,...,...,...,...,...
856,857,1,1,"Wick, Mrs. George Dennick (Mary Hitchcock)",female,45.0,1,1,36928,164.8667,NaN,S
862,863,1,1,"Swift, Mrs. Frederick Joel (Margaret Welles Ba...",female,48.0,0,0,17466,25.9292,D17,S
871,872,1,1,"Beckwith, Mrs. Richard Leonard (Sallie Monypeny)",female,47.0,1,1,11751,52.5542,D35,S
879,880,1,1,"Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)",female,56.0,0,1,11767,83.1583,C50,C


## Sort by Fare

In [5]:
top_fare = titanic.sort_values(by="Fare", ascending=False)

top_fare[["Name", "Fare"]].head(10)

,Name,Fare
679,"Cardeza, Mr. Thomas Drake Martinez",512.3292
258,"Ward, Miss. Anna",512.3292
737,"Lesurer, Mr. Gustave J",512.3292
88,"Fortune, Miss. Mabel Helen",263.0000
438,"Fortune, Mr. Mark",263.0000
341,"Fortune, Miss. Alice Elizabeth",263.0000
27,"Fortune, Mr. Charles Alexander",263.0000
742,"Ryerson, Miss. Susan Parker ""Suzette""",262.3750
311,"Ryerson, Miss. Emily Borie",262.3750
299,"Baxter, Mrs. James (Helene DeLaudeniere Chaput)",247.5208


## Create AgeGroup Column

In [6]:
def age_group(age):
    if pd.isna(age):
        return "Unknown"
    elif age < 13:
        return "Child"
    elif age <= 19:
        return "Teen"
    elif age <= 59:
        return "Adult"
    else:
        return "Senior"

titanic["AgeGroup"] = titanic["Age"].apply(age_group)

titanic[["Age", "AgeGroup"]].head(10)

,Age,AgeGroup
0,22.0,Adult
1,38.0,Adult
2,26.0,Adult
3,35.0,Adult
4,35.0,Adult
5,NaN,Unknown
6,54.0,Adult
7,2.0,Child
8,27.0,Adult
9,14.0,Teen


## Handle Missing Values

## Check Missing Values

In [7]:
titanic.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
AgeGroup         0
dtype: int64

## Fill Age with Median

In [8]:
titanic["Age"] = titanic["Age"].fillna(titanic["Age"].median())

## Fill Embarked with Mode

In [9]:
titanic["Embarked"] = titanic["Embarked"].fillna(
    titanic["Embarked"].mode()[0]
)

## Fill Cabin

In [10]:
titanic["Cabin"] = titanic["Cabin"].fillna("Unknown")

## Verify

In [11]:
titanic.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
AgeGroup       0
dtype: int64

## Remove Duplicates

In [12]:
duplicates = titanic.duplicated().sum()

print("Duplicates:", duplicates)

Duplicates: 0


## Group & Aggregate

## Survival Rate by Sex

In [13]:
survival_by_sex = titanic.groupby("Sex")["Survived"].mean()

print(survival_by_sex)

Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64


## Survival Rate by Passenger Class

In [14]:
survival_by_class = titanic.groupby("Pclass")["Survived"].mean()

print(survival_by_class)

Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64


## Pivot Table

In [15]:
pivot = titanic.pivot_table(
    values="Survived",
    index="Pclass",
    columns="Sex",
    aggfunc="mean"
)

pivot

Sex,female,male
Pclass,,
1,0.968085,0.368852
2,0.921053,0.157407
3,0.500000,0.135447


## Extract Passenger Title

In [17]:
titanic["Title"] = titanic["Name"].str.extract(r',\s*([^\.]+)\.')

titanic["Title"].value_counts()

Title
Mr              517
Miss            182
Mrs             125
Master           40
Dr                7
Rev               6
Major             2
Mlle              2
Col               2
Don               1
Mme               1
Ms                1
Lady              1
Sir               1
Capt              1
the Countess      1
Jonkheer          1
Name: count, dtype: int64

## Survival Rate by Title

In [18]:
title_survival = titanic.groupby("Title")["Survived"].mean().sort_values(ascending=False)

title_survival

Title
Lady            1.000000
Ms              1.000000
Sir             1.000000
Mme             1.000000
the Countess    1.000000
Mlle            1.000000
Mrs             0.792000
Miss            0.697802
Master          0.575000
Major           0.500000
Col             0.500000
Dr              0.428571
Mr              0.156673
Capt            0.000000
Jonkheer        0.000000
Don             0.000000
Rev             0.000000
Name: Survived, dtype: float64

In [20]:
titanic.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
AgeGroup       0
Title          0
dtype: int64

## Summary

- Dataset contains **891 passengers**.
- **Cabin** has the highest number of missing values.
- **Age** has a moderate amount of missing data.
- **Embarked** has only two missing values.
- Female passengers had a much higher survival rate than males.
- First-class passengers survived more often than third-class passengers.
- Passenger titles provided additional insights beyond gender (e.g., **Master**, **Mrs**, **Miss**, **Mr**).